03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,380 @@
+Use `?mapIds` to see more options or strategies.
+
+In the next chunk, we will run the `mapIds()` function and supply the `multiVals` argument with the `"list"` option in order to get a large list with all the mapped values found for each gene identifier.
+
+```{r}
+# Map Ensembl IDs to their associated Entrez IDs
+mapped_list <- mapIds(
+  org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+  keys = df$Gene,
+  column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+  multiVals = "list"
+)
+```
+
+## Explore gene ID conversion
+
+Now, let's take a look at our mapped object to see how the mapping went.
+
+```{r}
+# Let's use the `head()` function to take a preview at our mapped list
+head(mapped_list)
+```
+
+It looks like we have Entrez IDs that were successfully mapped to the Ensembl IDs we provided.
+However, the data is now in a `list` object, making it a little more difficult to explore.
+We are going to turn our list object into a data frame object in the next chunk.
+
+```{r}
+# Let's make our object a bit more manageable for exploration by turning it into a data frame
+mapped_df <- mapped_list %>%
+  tibble::enframe(name = "Ensembl", value = "Entrez") %>%
+  tidyr::unnest() # enframe makes a `list` column, so here we are converting that to simpler format
+```
+
+Now let's take a peek at our data frame.
+
+```{r}
+head(mapped_df)
+```
+
+We can see that our data frame has a new column `Entrez`.
+Let's get a summary of the values returned in the `Entrez` column of our mapped data frame.
+
+```{r}
+# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
+summary(as.factor(mapped_df$Entrez)) # We need to use the `as.factor()` function here in order to get the count of each unique value returned
+```
+
+There are 9102 NA's in our data frame, which means that 9102 out of the 31882 Ensembl IDs did not map to Entrez IDs.
+This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
+Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.
+
+Now let's check to see how many genes we have that were mapped to multiple IDs.
+
+```{r}
+multi_mapped_df <- mapped_df %>%
+  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
+  dplyr::count(Ensembl, name = "entrez_id_count") %>%
+  # Arrange by the genes with the highest number of Entrez IDs mapped
+  dplyr::arrange(desc(entrez_id_count))
+
+# Let's look at the first 6 rows of our `multi_mapped` object
+head(multi_mapped_df)
+```
+
+Looks like we have some cases where 15 Entrez IDs mapped to a single Ensembl ID.
```

Comment:
Keeping the exclamation point, but changing direction to match the way we are going...
```suggestion
Looks like we have one case where one Ensembl id mapped to 15 Entrez IDs!
```